### PR TITLE
Add push to Hub for PPOTrainer

### PR DESCRIPTION
**Pull request by lewtun**
This PR adds basic `push_to_hub()` functionality to the `PPOTrainer`. Here's a model produced from it: https://huggingface.co/lewtun/dummy-trl-model

There's a few questions I have about the desired functionality for end users:

* What is the expected usage for a trained `trl` model? The examples in the codebase are focused on training and I'm wondering if inference is simply a forward pass through the model with the value head, text-generation or something more complicated? 
* Should the reference model also be pushed to the Hub? For now, I've just pushed the active model with the value head

Here's a snippet that shows how this feature works:

```python
import torch
from transformers import AutoTokenizer
from trl import PPOTrainer, PPOConfig, AutoModelForCausalLMWithValueHead
from datasets import load_dataset
from trl.core import LengthSampler

config = PPOConfig(
    model_name="lvwerra/gpt2-imdb",
    learning_rate=1.41e-5,
)

def build_dataset(config, dataset_name="imdb", input_min_text_length=2, input_max_text_length=8):
    tokenizer = AutoTokenizer.from_pretrained(config.model_name)
    tokenizer.pad_token = tokenizer.eos_token
    ds = load_dataset(dataset_name, split='train')
    ds = ds.rename_columns({'text': 'review'})
    ds = ds.filter(lambda x: len(x["review"])>200, batched=False)
    input_size = LengthSampler(input_min_text_length, input_max_text_length)

    def tokenize(sample):
        sample["input_ids"] = tokenizer.encode(sample["review"])[:input_size()]
        sample["query"] = tokenizer.decode(sample["input_ids"])
        return sample

    ds = ds.map(tokenize, batched=False)
    ds.set_format(type='torch')
    return ds

dataset = build_dataset(config)

def collater(data):
    return dict((key, [d[key] for d in data]) for key in data[0])

model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name)
ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name)
tokenizer = AutoTokenizer.from_pretrained(config.model_name)
tokenizer.pad_token = tokenizer.eos_token

ppo_trainer = PPOTrainer(config, model, ref_model, tokenizer, dataset=dataset, data_collator=collater)
ppo_trainer.push_to_hub("dummy-trl-model")

# Load from Hub - note it retains the value-head
hub_model = AutoModelForCausalLMWithValueHead.from_pretrained("dummy-trl-model")
```

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thank you very much for working on this 💯 !
To the best of my knowledge once a the ppo trainer has been trained the only thing we want to retain is the trained model w/o its value head to perform any task you want with the remaining model that you can do with a standard `CausalLM` model. 
Note that currently you can also load a trained model with PPO using `transformers`:
```
model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name)
ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(config.model_name)
tokenizer = AutoTokenizer.from_pretrained(config.model_name)
tokenizer.pad_token = tokenizer.eos_token

ppo_trainer = PPOTrainer(config, model, ref_model, tokenizer, dataset=dataset, data_collator=collater)

# train ... then:
ppo_trainer.push_to_hub("dummy-trl-model")

# Load from Hub using transformers
from transformers import AutoModelForCausalLM

hub_model = AutoModelForCausalLM.from_pretrained("dummy-trl-model")
```
since `save_pretrained` from `AutoModelForCausalLMWithValueHead` is just a wrapper around the `save_pretrained` method from the base model: https://github.com/lvwerra/trl/blob/master/trl/models/modeling_base.py#L105-L115 

The only reason I see to share the `v_head`'s weights would be to continue a training. E.g.
```
from trl import PPOTrainer, AutoModelForCausalLMWithValueHead

model = AutoModelForCausalLMWithValueHead.from_pretrained("trl-gpt2-intermediate-training")
ppo_trainer = PPOTrainer(model, ...)
for i in range(epochs):
    ppo_trainer.step(...)

```
(I am unsure if this is what we want currently)

If this is what we want we would certainly change some logic of `AutoModelForCausalLMWithValueHead` class, make it maybe inherit from `transformers.PreTrainedModel` so that we can call load the `v_head` as well using `from_pretrained`, and probably add the class attribute `_keys_to_ignore_on_load_missing=["v_head.summary.weight", "v_head.summary.bias"]` so that we can initialize a `AutoModelForCausalLMWithValueHead` from any HF `CausalLM` model. 

For now I can see that the `v_head` might be not needed for the `ref_model` as the only place the forward pass from the `ref_model` is called is [here](https://github.com/lvwerra/trl/blob/4ec925552e4c3b15dc61547e29abee1204da5ca3/trl/trainer/ppo_trainer.py#L399).

Let's wait for @lvwerra and @edbeeching to chime in as I might be missing some details here !

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -41,6 +42,42 @@
...
+```
+
+You can then run inference as follows:
+
+```python
+from transformers import AutoTokenizer
+from trl import AutoModelForCausalLMWithValueHead
+
+tokenizer = AutoTokenizer.from_pretrained({model_name})
+model = AutoModelForCausalLMWithValueHead.from_pretrained({model_name})
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:73]: Currently this will load a new value head as the `save_pretrained` method of `AutoModelForCausalLMWithValueHead` does not save the value head. I think we should save it. I don't know if @younesbelkada wants to add the necessary changes to this or a follow-up PR.

Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:73]: Actually, to use the model you don't need TRL and you could just load it with transformers `AutoModelForCausalLM`. Installing TRL and the current example could be a second example if you want to continue training or get the value of the value head.

**Review by lvwerra — COMMENTED** [review]
Looks great, thanks! Left a comment regarding the value head. 

Before we merge could you add some tests??

File: trl/trainer/ppo_trainer.py (hunk 1 above)
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:73]: Yes I can take care of that in a follow-up PR

**Comment by lewtun** [comment]
Thanks for the comments @lvwerra ! I left a few questions that could do with your feedback - in the meantime I'll add some tests :)

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -41,6 +42,42 @@
 from trl.trainer.utils import AdaptiveKLController, FixedKLController
 
 
+MODEL_CARD_TEMPLATE = """---
+license: apache-2.0
+tags:
+- trl
+- transformers
+- reinforcement-learning
+- human-feedback
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:51]: maybe `human-feedback` should be added by the user manually, since you don't necessarily need to train from human feedback (e.g. all the sentiment examples).

File: trl/trainer/ppo_trainer.py [hunk 3]
```diff
@@ -41,6 +42,42 @@
...
+
+```bash
+python -m pip install trl
+```
+
+You can then run inference as follows:
+
+```python
+from transformers import AutoTokenizer
+from trl import AutoModelForCausalLMWithValueHead
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:70]: Both `AutoModelForCausalLMWithValueHead` (`trl`, used for training) and `AutoModelForCausalLM` (`transformers`, used for inference) can be used. in the example the latter is enough I would say.

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -41,6 +42,42 @@
...
+
+tokenizer = AutoTokenizer.from_pretrained({model_name})
+model = AutoModelForCausalLMWithValueHead.from_pretrained({model_name})
+
+inputs = tokenizer("Hello, my llama is cute", return_tensors="pt")
+outputs = model(**inputs, labels=inputs["input_ids"])
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:76]: you could also pass it to a `text-generation` pipeline

File: trl/trainer/ppo_trainer.py [hunk 5]
```diff
@@ -658,3 +695,22 @@ def log_stats(
 
                 dist.barrier()
                 dist.all_reduce(rewards, op=torch.distributed.ReduceOp.SUM)
+
+    def create_model_card(self, path: str, model_name: Optional[str] = "TRL Model") -> None:
```
[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:699]: Sounds good, we could potentially also use the same method since `transformers` is already a dependency?

File: trl/trainer/ppo_trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/ppo_trainer.py:83]: Done in https://github.com/lvwerra/trl/pull/68/commits/47289f5bf3b20df8a5870478dc68fd6032af20ad

File: trl/trainer/ppo_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/ppo_trainer.py:77]: Sounds good!

File: trl/trainer/ppo_trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/ppo_trainer.py]: Fixed in https://github.com/lvwerra/trl/pull/68/commits/47289f5bf3b20df8a5870478dc68fd6032af20ad

File: trl/trainer/ppo_trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lewtun) [reply trl/trainer/ppo_trainer.py:817]: Unfortunately, the `create_model_card` in `transformers` is a method of the `Trainer`, so can't be re-used. I suggest we just keep it simple for now and see if there's a wish from the community to expand it with more tags etc.

File: trl/trainer/ppo_trainer.py (hunk 1 above)
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:80]: it has been taken care in https://github.com/lvwerra/trl/pull/86 !

File: tests/trainer/test_ppo_trainer.py [hunk 6]
```diff
@@ -61,6 +67,37 @@ class PPOTrainerTester(unittest.TestCase):
...
+            except HTTPError:
+                pass
+
+    def setUp(self):
+
+        # model_id
+        model_id = "gpt2"
+
+        # get models and tokenizer
+        self.gpt2_model = AutoModelForCausalLMWithValueHead.from_pretrained(model_id)
```
Comment (lewtun) [inline tests/trainer/test_ppo_trainer.py:91]: Do we need fresh models for each unit test? If not, I can move this to `setUpClass()` which is only called once (and saves time)

**Review by lewtun — COMMENTED** [review]
@lvwerra @younesbelkada this PR looks good to go on my side - note that I refactored the test filenames to reflect the repo structure. This is what we do in `transformers` and I personally find it to be a good organising principle as the project grows. Happy to revert if you wish!

File: tests/trainer/test_ppo_trainer.py (hunk 6 above)
Comment (younesbelkada) [reply tests/trainer/test_ppo_trainer.py:91]: Actually I don't think we need a fresh model for each test, yes can you move it to `setUpClass()` instead 🙏 Thanks!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for shipping this! 💯 
Let's just refactor the test as you proposed in https://github.com/lvwerra/trl/pull/68#discussion_r1084093252 otherwise LGTM

File: tests/trainer/test_ppo_trainer.py (hunk 6 above)
Comment (lewtun) [reply tests/trainer/test_ppo_trainer.py]: Done in https://github.com/lvwerra/trl/pull/68/commits/6c4ab687c2b9c24bee4fe9a56aa5877e1f8b896c

**Review by lvwerra — APPROVED** [review]
Looks good to me - thanks for adding!

File: trl/trainer/ppo_trainer.py [hunk 7]
```diff
@@ -793,3 +839,23 @@ def log_stats(
...
+        """Creates and saves a model card for a TRL model.
+
+        Args:
+            path (`str`): The path to save the model card to.
+            model_name (`str`, *optional*): The name of the model, defaults to `TRL Model`.
+        """
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+        user = whoami()["name"]
```
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:853]: This line is the reason why the unit test in the staging endpoint fails - by default `whoami()` points to hf.co, when we need it to point to https://hub-ci.huggingface.co

One can do this by using `HfApi(endpoint="https://hub-ci.huggingface.co")`, but then we need to set an env var for `endpoint` which feels clunky. 

For now, I propose we skip the unit test until we find a better solution to test `push_to_hub` features

OK with you @younesbelkada @lvwerra ?

Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:853]: One way to solve this is by patching the function in the test? we could have a dummy function for whoami that returns whatever we want during the testing? 

happy to merge it as is and figure it out later :)

Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:853]: same as @lvwerra , definitely OK for me to skip this test for now and look at it later!
